hs_core/views/utils.py

```diff
@@ -559,84 +559,125 @@ def get_metadata_contenttypes():
-            is_favorite=Case(When(short_id__in=favorite_resources.values_list('short_id', flat=True),
-                                  then=Value(True, BooleanField()))))
+def get_my_resources_list(user, annotate=False, filter=None, **kwargs):
+    """
+    Gets a QuerySet object for listing resources that belong to a given user.
+    :param user: an instance of User - user who wants to see his/her resources
+    :param annotate: whether to annotate for my resources page listing.
+    :return: an instance of QuerySet of resources
+    """
+    resource_collection = BaseResource.objects.none()
+    owned_resources = BaseResource.objects.none()
+    editable_resources = BaseResource.objects.none()
+    viewable_resources = BaseResource.objects.none()
+    discovered_resources = BaseResource.objects.none()
+
+    # if annotate:  # When used in the My Resources page, annotate for speed
+    if not filter or 'owned' in filter:
+        # get a list of resources with effective OWNER privilege
+        owned_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.OWNER)
+        # remove obsoleted resources from the owned_resources
+        owned_resources = owned_resources.exclude(object_id__in=Relation.objects.filter(
+            type='isReplacedBy').values('object_id')).exclude(extra_data__to_be_deleted__isnull=False)
+
+    if not filter or 'editable' in filter:
+        # get a list of resources with effective CHANGE privilege (should include resources that the
+        # user has access to via group
+        editable_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.CHANGE,
+                                                                             via_group=True)
+        # remove obsoleted resources from the editable_resources
+        editable_resources = editable_resources.exclude(object_id__in=Relation.objects.filter(
+            type='isReplacedBy').values('object_id'))
+
+    if not filter or 'viewable' in filter:
+        # get a list of resources with effective VIEW privilege (should include resources that the
+        # user has access via group
+        viewable_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.VIEW,
+                                                                             via_group=True)
+        # remove obsoleted resources from the viewable_resources
+        viewable_resources = viewable_resources.exclude(object_id__in=Relation.objects.filter(
+            type='isReplacedBy').values('object_id'))
+
+    if not filter or 'discovered' in filter:
+        discovered_resources = user.ulabels.my_resources
+        if annotate:
+            discovered_resources = discovered_resources.all().annotate(discovered=Value(True, BooleanField()))
+
+    favorite_resources = user.ulabels.favorited_resources
+
+    # join all queryset objects. if/else saves one database query
+    if not filter or 'is_favorite' in filter:
```

Comment:
@pkdash I _think (?)_ that not including the favorites unless they are requested helps slightly maybe with performance? If I'm wrong, we can remove the logic...

Reply:
@devincowan It seems the queryset (resource_collection) that we generate as part the get_my_resources_list() results in 2 queries per resource. Which means if we have to render 100 resources, the queryset resource_collection will execute 200 queries. So unless the user has more than 10 resources marked as favorite the 'is_favorite' filter may not make much difference. Earlier this week I was able to figure out a way to count the number of queries generated by get_my_resources_list() and the my_resources.html page. It seems real bottleneck is at the template. The template is generating minimum of 5 queries per resource.  That means in total minimum of 7 queries for one resource.  The template will generate 5 queries if the resource has one creator. For any additional creator that's a hydroshare user, the template will generate one more query. I think we may be able to fix it so that the template generates only 2 or 3 queries per resource independent of number of creators for the resource.

Reply:
Noting here, I will push 2 modifications on this PR:
1. `{{res.get_absolute_url}}` - generates 2 queries. The get_absolute_url() comes from the Page class. Will overwrite in AbstractResource class so that no query is needed.
2. `{{res|resource_first_author}}` - generates 1 query. Will modify the template tag

The last optimization we will deal with separately:
https://github.com/hydroshare/hydroshare/issues/4732
 {% if creator.relative_uri and creator.is_active %} - The call to is_active generates 1 query.